snooty/test_parser.py

```diff
@@ -140,73 +141,165 @@ def test_literalinclude() -> None:
+   :start-after: start example 1
+   :end-before: end example 1
+   :dedent: 4
    :linenos:
+   :copyable: false
 """,
     )
     page.finish(diagnostics)
     assert diagnostics == []
     check_ast_testing_string(
         page.ast,
         """<root>
-<code lang="py" copyable="True" linenos="True">db.inventory.insert_many([
-    {"item": "journal",
-     "qty": 25,
-     "tags": ["blank", "red"],
-     "size": {"h": 14, "w": 21, "uom": "cm"}},\n
-    {"item": "mat",
-     "qty": 85,
-     "tags": ["gray"],
-     "size": {"h": 27.9, "w": 35.5, "uom": "cm"}},\n
-    {"item": "mousepad",
-     "qty": 25,
-     "tags": ["gel", "blue"],
-     "size": {"h": 19, "w": 22.85, "uom": "cm"}}])</code>
-</root>""",
+        <directive name="literalinclude" copyable="False" dedent="4" linenos="True" end-before="end example 1" language="python" start-after="start example 1">
+        <text>/test_parser/includes/sample_code.py</text>
+        <code lang="python" linenos="True">print("test dedent")</code>
+        </directive>
+        </root>""",
     )
 
-    # Test bad code-blocks
+    # Test failure to specify argument file
     page, diagnostics = parse_rst(
         parser,
         path,
         """
-.. literalinclude:: /driver-examples/pythonexample.py
-   :start-after: Start Example 0
-   :end-before: End Example 3
+.. literalinclude::
 """,
     )
     page.finish(diagnostics)
-    assert len(diagnostics) == 1 and isinstance(diagnostics[0], InvalidLiteralInclude)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], ExpectedPathArg)
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <directive name="literalinclude">
+        </directive>
+        </root>""",
+    )
 
+    # Test failure to locate included code file
     page, diagnostics = parse_rst(
         parser,
         path,
         """
-.. literalinclude:: /driver-examples/pythonexample.py
-   :start-after: Start Example 3
-   :end-before: End Example 0
+.. literalinclude:: includes/does_not_exist.js
 """,
     )
     page.finish(diagnostics)
-    assert len(diagnostics) == 1 and isinstance(diagnostics[0], InvalidLiteralInclude)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], CannotOpenFile)
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <directive name="literalinclude">
+        <text>includes/does_not_exist.js</text>
+        </directive>
+        </root>""",
+    )
 
+    # Test failure to locate start-after text
     page, diagnostics = parse_rst(
         parser,
         path,
         """
-.. literalinclude:: /driver-examples/garbagnrekvjisd.py
+.. literalinclude:: /test_parser/includes/sample_code.py
+   :start-after: does not exist
+   :end-before: end example 1
 """,
     )
     page.finish(diagnostics)
-    assert len(diagnostics) == 1 and isinstance(diagnostics[0], CannotOpenFile)
+    assert len(diagnostics) == 1
+    assert isinstance(diagnostics[0], InvalidLiteralInclude)
```

Comment:
I like the handling of this error a lot—acting as if there is no `start-after` specified and still applying the rest of the options 👍 Perfect!

Reply:
I've updated `end_before` functionality to "match". If no `end_before_text` is found, it defaults to the EOF.